File: knapsack.py
def knapsack(cap, weights, n):
    if (n == 0) or (cap == 0):
        return 0

    evaluating = weights[n - 1]
    if evaluating > cap:
        return knapsack(cap, weights, n - 1)

    prev_result = evaluating + knapsack(cap - evaluating, weights, n - 1)
    result = knapsack(cap, weights, n - 1)
    return max(prev_result, result)


gallon_size = 5
bottles = [1.0, 3.0, 3.5, 1.5] * 60
print("initialized")
weights = bottles
cap = gallon_size
n = len(bottles)

result = knapsack(cap, weights, n)
print(result)

File: knapsack.rs
//my first rust code \o/

fn main() {
    let gallon_size: f64 = 5.0;
    let real_bottles: [f64; 4] = [1.0, 3.0, 3.5, 1.5];

    let bottles: Vec<f64> = (0..4*60)
        .map(|n| real_bottles[n % 4])
        .collect();

    let result = knapsack(gallon_size, &bottles, bottles.len());
    println!("{}", result);
}

fn knapsack(cap:f64, weights: &[f64], n:usize) -> f64 {
    if n == 0 || cap == 0.0 {
        return 0.0;
    }

    let evaluating = *(weights.get(n - 1).unwrap()) as f64;
    if evaluating > cap {
        return knapsack(cap, weights, n - 1);
    }

    let prev_result = evaluating + knapsack(cap - evaluating, weights, n - 1);
    let result = knapsack(cap, weights, n - 1);
    return prev_result.max(result);
}

